### PR TITLE
LF-30: add JsonLfReader and unit tests for happy cases

### PR DESCRIPTION
**Pull request by raphael-speyer-da**
Provide initial implementation of `JsonLfReader`, which can be used by Java code-gen to build the Java objects from a JSON LF formatted string, as specified here https://docs.daml.com/json-api/lf-value-specification.html

Not done yet:
* the actual code-gen to produces the `fromJson` methods on custom types
* unit tests for failure cases on parsing
* some corner-cases of rounding
* a `JsonLfWriter`

### CODE REVIEW

File: language-support/java/bindings/src/test/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReaderTest.java [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+        eq("[42]", Optional.of(Optional.of(42L))));
+  }
+
+  @Test
+  void testOptionalNestedDeeper() throws IOException {
+    checkReadAll(
+        r -> r.optionalNested(r.optionalNested(r.optional(r.int64()))),
+        eq("null", Optional.empty()),
+        eq("[]", Optional.of(Optional.empty())),
+        eq("[[]]", Optional.of(Optional.of(Optional.empty()))),
```
Comment (lucianojoublanc-da) [inline language-support/java/bindings/src/test/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReaderTest.java:170]: Unit tests look really comprehensive. Maybe add one for this nested list case when fields are non-optional?

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+        map.put(key, val);
+      }
+      ;
+      readEndObject();
+      return map;
+    };
+  }
+
+  // Read a map with unknown number of items of the same types.
+  public <K, V> FromJson<Map<K, V>> genMap(FromJson<K> readKey, FromJson<V> readValue) {
```
Comment (lucianojoublanc-da) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java:206]: How about including a convenience instance for `Set`? This is really unpleasant to deal with manually. i.e. I mean the horrible
```json
{
  "a" : {},
  "b" : {}
} === Set("a", "b")
```

**Review by lucianojoublanc-da — COMMENTED** [review]
Looking forward to the version with codegen, so we can test it out!

File: language-support/java/bindings/src/test/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReaderTest.java (hunk 1 above)
Comment (raphael-speyer-da) [reply language-support/java/bindings/src/test/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReaderTest.java:170]: I added one for `Optional[List[List[Int64]]]`

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java (hunk 2 above)
Comment (raphael-speyer-da) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java:206]: Once code gen exists, the idea is that decoding sets would reduce to `Set.fromJson(json)`, where `Set` is the code-gen version of the one in the daml stdlib.

If we wanted native support for `Set`, beyond the daml library implementation, I think we'd want to make it a concept in the json lf encoding. There's also always the option of convenience functions to convert from the code-gen Set and a `java.util.Set` and/or `scala.collection.immutable.Set`

**Comment by raphael-speyer-da** [comment]
@stephencompall-DA as discussed, that plan may be worthwhile at some point, but obviously involves translating existing Scala logic and libraries to Java, as well as converting through the intermediate representations at runtime. The interface that a client uses needn't be affected if we did want to change the implementation.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/FromJson.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+// to tell them how to read that argument type.
+//
+// e.g.
+//   String str = reader.text().read();
+// or
+//   List<String> = reader.list(reader::text).read();
+public interface FromJson<T> {
+  public T read() throws Error;
+
+  public class Error extends IOException {
```
Comment (cocreature) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/FromJson.java:19]: should this be static?

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+        map.put(key, val);
+      }
+      ;
+      readEndArray();
+      return map;
+    };
+  }
+
+  // The T type should not itself be Optional<?>. In that case use OptionalNested below.
+  public <T> FromJson<Optional<T>> optional(FromJson<T> readValue) {
```
Comment (cocreature) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java:225]: is it worth adding a runtime typecheck that T is not an instance of Optional?

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+        for (String fieldName : fieldNames) {
+          var field = fieldsByName.apply(fieldName);
+          args[field.argIndex] = field.fromJson.read();
+        }
+        readEndArray();
+      } else {
+        parseExpected("object or array");
+      }
+
+      // Handle missing and unknown fields.
```
Comment (cocreature) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java:330]: generally JSON decoders tend to be lax wrt to unknown fields. This also fits in nicely with the new upgrading proposal. I would suggest to be lax here from the start.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+      }
+      return null;
+    };
+  }
+
+  // Provides a generic way to read a variant type, by specifying each tag.
+  public <T> FromJson<T> variant(List<String> tagNames, TagReader<T> readTag) {
+    return () -> {
+      readStartObject();
+      if (!readFieldName().equals("tag")) parseExpected("tag field");
```
Comment (cocreature) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java:266]: I might be misunderstanding how jackson works here but to me this looks like you're making assumptions on the order of keys. Specifically you seem to assume that "tag" always comes before "value". Your tests also only cover this case. That doesn't seem right to me. JSON objects are generally unordered an a lot of encoders don't even give you control over the field order.

**Review by cocreature — APPROVED** [review]
Nice work! Please add a test for field ordering on variants and fix it if it breaks before merging.

**Comment by stephencompall-DA** [comment]
> but obviously involves translating existing Scala logic and libraries to Java

Since `ledger-service/lf-value-json` didn't entail any such translation, I do not believe that plan entails any further such translation.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java (hunk 4 above)
Comment (raphael-speyer-da) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java:225]: Sure. If this happened, the bug would be in the code-gen, so they're an extra step removed from being able to handle it, but at least we can report it in a useful way.

I'd actually like to come back to this to see whether we can avoid having these two different decoders but this was what I got working initially.

But in the meantime, done.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java (hunk 5 above)
Comment (cocreature) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java:337]: seems fine to fail for now and revisit this once we get to the new upgrading proposal.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java (hunk 6 above)
Comment (cocreature) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java:273]: I'd suggest we fix it before merging.

Comment (raphael-speyer-da) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java]: Done
